File: code_room/app/channels/doc_channel.rb
class DocChannel < ApplicationCable::Channel
  def subscribed
    stream_from "doc_channel_#{params[:document_id]}"
    if params[:editing]
      doc_connection_params = { 
        editor_id: connection.current_user.id, 
        document_id: params[:document_id]
      }
      
      DocumentConnection.where(doc_connection_params).destroy_all
      @doc_connection = DocumentConnection.create(doc_connection_params)
      send_initial_state
    end

    broadcast_active_editors
  end

  def receive(data)
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", data)
  end

  def unsubscribed
    if @doc_connection
      @doc_connection.destroy
    else
      doc_connection_params = { 
        editor_id: connection.current_user.id, 
        document_id: params[:document_id]
      }
      
      DocumentConnection.where(doc_connection_params).destroy_all
    end
    
    
    broadcast_active_editors if params[:editing]
  end

  def broadcast_active_editors
    editors = DocumentConnection.where(document_id: params[:document_id])
    .map do |conn|
      User.find(conn.editor_id)
      .attributes
      .slice('id', 'username', 'avatar_url', 'email')
    end

    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", {editors: editors})
  end

  private

  def send_initial_state
    if DocumentConnection.where(document_id: params[:document_id]).length == 1
      doc = Document.find(params[:document_id])
      file_name = doc.file_name
      content = doc.content
      document = { content: content, file_name: file_name }

      ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", { document: document })
      return
    end

    syncRequest = { sender_id: connection.current_user.id, sync: true }
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", syncRequest)
  end
end

File: code_room/app/javascript/channels/doc_channel.js
import consumer from "./consumer"

   const connectToDoc = (docId, editing, callbacks, userId) => {
    return consumer.subscriptions.create({
      channel: "DocChannel", 
      document_id: docId,
      editing
    }, {
      connected() {
        console.log("doc channel connected...")
        if (editing) callbacks.connect();
      },

      disconnected() {
        console.log("doc channel disconnected")
        // Called when the subscription has been terminated by the server
      },

      received(data) {
        if (data.join || data.offer || data.iceCandidate || data.answer) {
          if (data.senderId === userId || !editing) return;
          callbacks.initConnection(data)
          return
        }
        
        if (data.saved_state) {
          callbacks.save(data)
          return;
        }
        
        if (data.sync) {
          callbacks.sendState(data);
          return;
        }

        if(data.currentState) {
          callbacks.syncState(data);
          return;
        }
        
        if (data.document) {
          callbacks.initialize(data)
          return;
        }
        
        if (data.editors) {
          callbacks.editorList(data);
          return;
        }
        
        data.backup ? callbacks.edit(data) : callbacks.cursor(data)
      },

      resubscribe() {
        this.perform("subscribed");
      }
    });
  }


  export default connectToDoc;